README
ocaml-glpk - OCaml bindings for glpk
====================================

Author: Samuel Mimram <samuel.mimram@ens-lyon.org>
Email: smimram@users.sourceforge.net
Homepage: http://ocaml-glpk.sourceforge.net/

Copyright (C) 2003-2010 Samuel Mimram.


Dependencies
============
To build this library you need to have OCaml 3.07 and glpk installed.

You can download glpk here: http://www.gnu.org/software/glpk/glpk.html.


Installation
============
To compile the program type:

make

then, to install it, type as root:

make install


Notes
=====
The bindings are not (yet) complete. If you need a specific function, just mail
me (or send a patch) and I'll add it as soon as possible.


License
=======
The license of this library is the GPL like the original glpk library. It is
therefore useless to ask me to change the license, I can't do that.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.


CHANGES
0.2.0 (unreleased)
=====
* Adapt to new glpk API, thus many backward incompatible changes, sorry.
* Switch build system to dune.

0.1.8 (2016-11-13)
=====
* Remove -ansi flag in order to be compatible with OCaml 4.04.
* Remove lpx_main from oldapi since glp_main has been removed in glpk.

0.1.7 (2014-07-31)
=====
* Use compatibility layer for old API which was removed in glpk 4.53.

0.1.6 (2010-07-02)
=====
* Added load_sparse_matrix.
* Added branch_and_bound_opt.
* Added get_row_primal and get_row_dual (thanks Sean McLaughlin).
* Protect values used in blocking sections.

0.1.5 (2006-07-25)
=====
* Added set/get_simplex_iteration_count.
* Fixed a bug in new_blp (unnecessary CAMLlocal) and removed unnecessary
  CAMLparams (thanks Markus Mottl).
* Allow other threads to run while simplex, interior or branch_and_bound is
  being executed (added blocking sections).

0.1.4 (2006-04-07)
=====
* Corrected get_obj_val to handle mip problems.
* get_col_primal now return correct values for mip problems.
* Added {read,write}_cplex functions.

0.1.3 (2005-09-24)
=====
* Corrected forgotten indice shifting in set_col_kind (thanks Pierre Lemaire).
* Corrected the get_col_primals function, the returned array's length is now the
  number of columns of the lp (found by Erick Matsen).
* Added the get_num_cols function.
* Using caml_alloc_custom in ocaml_glpk_new_prob, it should fix a possible
  memory corruption problem (reported by Stefan Ratschan).

0.1.2 (2005-04-18)
=====
* Put Val_int instead of Int_val in ocaml_glpk_get_direction and
  ocaml_glpk_get_class and added a unit parameter to ocaml_glpk_new_prob
  (thanks Mike Furr).
* Corrected a bug in load_matrix: glpk only wants non-null elements (found by
  Jeremy Dubreil).
* Added the use_presolver function to get exception with simplex when there is
  no feasible solution.
* Using caml namespace functions.

0.1.1 (2004-06-23)
=====
* Added functions: get_prob_name, get_obj_name, get_class, get_direction,
  get_row_name, get_col_name, warm_up.
* We are now raising exceptions on some errors (they are not all handled yet).

0.1.0 (2004-04-24)
=====
* Initial release.


src/glpk.ml
(*
 * ocaml-glpk - OCaml bindings to glpk
 * Copyright (C) 2004-2019 Samuel Mimram
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *)

type lp

type direction = Minimize | Maximize

type aux_var_type = Free_var | Lower_bounded_var | Upper_bounded_var | Double_bounded_var | Fixed_var

type prob_class = Linear_prog | Mixed_integer_prog

type var_kind = Continuous_var | Integer_var | Boolean_var

type status = Optimal | Feasible | Infeasible | No_feasible | Unbounded | Undefined

type message_level = Msg_off | Msg_err | Msg_on | Msg_all

exception Time_limit
exception Unknown_error of int

let () =
  Callback.register "ocaml_glpk_exn_time_limit" Time_limit;
  Callback.register_exception "ocaml_glpk_exn_unknown" (Unknown_error 0)

external new_problem : unit -> lp = "ocaml_glpk_new_prob"

external set_prob_name : lp -> string -> unit = "ocaml_glpk_set_prob_name"

external get_prob_name : lp -> string = "ocaml_glpk_get_prob_name"

external set_obj_name : lp -> string -> unit = "ocaml_glpk_set_obj_name"

external get_obj_name : lp -> string = "ocaml_glpk_get_obj_name"

external set_direction : lp -> direction -> unit = "ocaml_glpk_set_direction"

external get_direction : lp -> direction = "ocaml_glpk_get_direction"

external add_rows : lp -> int -> unit = "ocaml_glpk_add_rows"

external set_row_name : lp -> int -> string -> unit = "ocaml_glpk_set_row_name"

external get_row_name : lp -> int -> string = "ocaml_glpk_get_row_name"

external set_row_bounds : lp -> int -> aux_var_type -> float -> float -> unit = "ocaml_glpk_set_row_bounds"

external add_columns : lp -> int -> unit = "ocaml_glpk_add_cols"

external set_col_name : lp -> int -> string -> unit = "ocaml_glpk_set_col_name"

external get_col_name : lp -> int -> string = "ocaml_glpk_get_col_name"

external set_col_bounds : lp -> int -> aux_var_type -> float -> float -> unit = "ocaml_glpk_set_col_bounds"

external set_obj_coef : lp -> int -> float -> unit = "ocaml_glpk_set_obj_coef"

external load_matrix : lp -> float array array -> unit = "ocaml_glpk_load_matrix"

external load_sparse_matrix : lp -> ((int * int) * float) array -> unit = "ocaml_glpk_load_sparse_matrix"

type simplex_param =
  {
    message_level : message_level option;
    time_limit : int option;
  }

external simplex : lp -> simplex_param -> unit = "ocaml_glpk_simplex"

let simplex ?message_level ?time_limit lp =
  let p = { message_level; time_limit } in
  simplex lp p

external get_obj_val : lp -> float = "ocaml_glpk_get_obj_val"

external get_col_primal : lp -> int -> float = "ocaml_glpk_get_col_prim"

external get_row_primal : lp -> int -> float = "ocaml_glpk_get_row_prim"

external get_row_dual : lp -> int -> float = "ocaml_glpk_get_row_dual"

let make_problem dir zcoefs constr pbounds xbounds =
  let lp = new_problem () in
    set_direction lp dir;
    add_rows lp (Array.length pbounds);
    for i = 0 to (Array.length pbounds) - 1 do
      match pbounds.(i) with
        | lb, ub when lb = -.infinity && ub = infinity -> set_row_bounds lp i Free_var 0. 0.
        | lb, ub when ub = infinity -> set_row_bounds lp i Lower_bounded_var lb 0.
        | lb, ub when lb = -.infinity -> set_row_bounds lp i Upper_bounded_var 0. ub
        | lb, ub when lb = ub -> set_row_bounds lp i Fixed_var lb ub
        | lb, ub -> set_row_bounds lp i Double_bounded_var lb ub
    done;
    add_columns lp (Array.length xbounds);
    for i = 0 to (Array.length xbounds) - 1 do
      set_obj_coef lp i zcoefs.(i);
      match xbounds.(i) with
        | lb, ub when lb = -.infinity && ub = infinity -> set_col_bounds lp i Free_var 0. 0.
        | lb, ub when ub = infinity -> set_col_bounds lp i Lower_bounded_var lb 0.
        | lb, ub when lb = -.infinity -> set_col_bounds lp i Upper_bounded_var 0. ub
        | lb, ub when lb = ub -> set_col_bounds lp i Fixed_var lb ub
        | lb, ub -> set_col_bounds lp i Double_bounded_var lb ub
    done;
    load_matrix lp constr;
    lp

external get_num_rows : lp -> int = "ocaml_glpk_get_num_rows"

external get_num_cols : lp -> int = "ocaml_glpk_get_num_cols"

let get_col_primals lp =
  let n = get_num_cols lp in
  let ans = Array.make n 0. in
    for i = 0 to (n - 1)
    do
      ans.(i) <- get_col_primal lp i
    done;
    ans

external scale_problem : lp -> unit = "ocaml_glpk_scale_problem"

external unscale_problem : lp -> unit = "ocaml_glpk_unscale_problem"

external interior : lp -> unit = "ocaml_glpk_interior"

external set_col_kind : lp -> int -> var_kind -> unit = "ocaml_glpk_set_col_kind"

external warm_up : lp -> unit = "ocaml_glpk_warm_up"

type intopt_param =
  {
    message_level : message_level option;
    time_limit : int option;
  }

external branch_and_cut : lp -> intopt_param -> unit = "ocaml_glpk_intopt"

let branch_and_cut ?message_level ?time_limit lp =
  let p = { message_level; time_limit } in
  branch_and_cut lp p

external mip_status : lp -> status = "ocaml_glpk_mip_status"

external mip_obj_val : lp -> float = "ocaml_glpk_mip_obj_val"

external mip_row_val : lp -> int -> float = "ocaml_glpk_mip_row_val"

external mip_col_val : lp -> int -> float = "ocaml_glpk_mip_col_val"

external read_cplex : string -> lp = "ocaml_glpk_read_cplex"

external write_cplex : lp -> string -> unit = "ocaml_glpk_write_cplex"
